src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs

```diff
@@ -68,21 +69,24 @@ private async Task AnalyzeDocumentForKindAsync(TextDocument document, AnalysisKi
 
                 // First attempt to fetch diagnostics from the cache, while computing the analyzers that are not cached.
                 using var _ = ArrayBuilder<(DiagnosticAnalyzer analyzer, ActiveFileState state)>.GetInstance(out var nonCachedAnalyzersAndStates);
+                using var argsBuilder = TemporaryArray<DiagnosticsUpdatedArgs>.Empty;
                 foreach (var stateSet in stateSets)
                 {
                     var (activeFileState, existingData) = TryGetCachedDocumentAnalysisData(document, stateSet, kind, version,
                         backgroundAnalysisScope, compilerDiagnosticsScope, isActiveDocument, isVisibleDocument,
                         isOpenDocument, isGeneratedRazorDocument, cancellationToken, out var isAnalyzerSuppressed);
                     if (existingData.HasValue)
                     {
-                        PersistAndRaiseDiagnosticsIfNeeded(existingData.Value, stateSet.Analyzer, activeFileState);
+                        PersistAndAddDiagnosticsArgsIfNeeded(ref argsBuilder.AsRef(), existingData.Value, stateSet.Analyzer, activeFileState);
                     }
                     else if (!isAnalyzerSuppressed)
                     {
                         nonCachedAnalyzersAndStates.Add((stateSet.Analyzer, activeFileState));
                     }
                 }
 
+                AnalyzerService.RaiseDiagnosticsUpdated(argsBuilder.ToImmutableAndClear());
```

Comment:
Consider adding a comment here stating that we raise DiagnosticsUpdated event in a batch for all cached analyzers as we expect all the operations in the prior loop to have minimal performance cost. This is different from the subsequent loop that executes non-cached analyzers, where it can have a user observable delay, especially for compiler analyzer which needs to be executed upfront and its DiagnosticsUpdated event raised as soon as possible.

Reply:
➡️ Added comment